lib/service_provider.rb

```diff
@@ -1,105 +1,63 @@
 require 'fingerprinter'
 
 class ServiceProvider
-  # currently acceptable encryption values are 'none' and 'aes256-cbc'
-  DEFAULT_ENCRYPTION = 'aes256-cbc'.freeze
-
   attr_reader :issuer
 
   def initialize(issuer)
     @issuer = issuer
   end
 
   def metadata
-    ignored_methods = [:metadata, :encrypt_responses?, :encryption_opts, :issuer]
-    metadata_methods = self.class.instance_methods(false) - ignored_methods
-
-    @metadata ||= metadata_methods.inject({}) do |hash, method|
-      hash.merge(:"#{method}" => send(method))
-    end
-  end
-
-  def acs_url
-    sp_attributes['acs_url']
-  end
-
-  def assertion_consumer_logout_service_url
-    sp_attributes['assertion_consumer_logout_service_url']
-  end
-
-  def sp_initiated_login_url
-    sp_attributes['sp_initiated_login_url']
-  end
-
-  def metadata_url
-    sp_attributes['metadata_url']
+    sp_attributes.merge(shared_attributes)
   end
 
-  def agency
-    sp_attributes['agency']
-  end
-
-  def friendly_name
-    sp_attributes['friendly_name']
-  end
-
-  def attribute_bundle
-    sp_attributes['attribute_bundle']
+  def encryption_opts
+    return nil unless encrypt_responses?
+    {
+      cert: ssl_cert,
+      block_encryption: block_encryption,
+      key_transport: 'rsa-oaep-mgf1p'
+    }
   end
 
-  def cert
-    sp_cert = sp_attributes['cert']
-    return if sp_cert.blank?
-
-    cert_dir = "#{Rails.root}/certs/sp/"
-
-    @cert ||= File.read("#{cert_dir}#{sp_cert}.crt")
-  end
+  private
 
-  def block_encryption
-    sp_attributes['block_encryption'] || DEFAULT_ENCRYPTION
+  def sp_attributes
+    @sp_attributes ||= config.sp_attributes
   end
 
-  def key_transport
-    'rsa-oaep-mgf1p'
+  def config
+    ServiceProviderConfig.new(filename: 'service_providers.yml', issuer: issuer)
   end
 
-  def encryption_opts
-    return nil unless encrypt_responses?
+  def shared_attributes
```

Comment:
I'm getting rid of the method altogether and just merging the fingerprint directly into `sp_attributes`. No need to use a separate method and try to come up with a good name when there's only one attribute to add.